# Customer Churn Prediction

## Dataset Understanding

### Objective

The objective of this project is to predict whether a customer is likely to leave a telecom company based on customer demographics, account information, and service usage.

This notebook focuses on understanding the dataset structure, identifying data quality issues, and preparing for exploratory data analysis.

In [5]:
import pandas as pd
import numpy as np

pd.set_option('display.max_columns', None)

In [ ]:
#Load dataset
df = pd.read_csv("../data/raw/Telco-Customer-Churn.csv")

## First Look at the Dataset

We begin by inspecting the first rows to understand the available features and their format.

In [8]:
df.head()

,customerID,gender,SeniorCitizen,Partner,Dependents,tenure,PhoneService,MultipleLines,InternetService,OnlineSecurity,OnlineBackup,DeviceProtection,TechSupport,StreamingTV,StreamingMovies,Contract,PaperlessBilling,PaymentMethod,MonthlyCharges,TotalCharges,Churn
0,7590-VHVEG,Female,0,Yes,No,1,No,No phone service,DSL,No,Yes,No,No,No,No,Month-to-month,Yes,Electronic check,29.85,29.85,No
1,5575-GNVDE,Male,0,No,No,34,Yes,No,DSL,Yes,No,Yes,No,No,No,One year,No,Mailed check,56.95,1889.5,No
2,3668-QPYBK,Male,0,No,No,2,Yes,No,DSL,Yes,Yes,No,No,No,No,Month-to-month,Yes,Mailed check,53.85,108.15,Yes
3,7795-CFOCW,Male,0,No,No,45,No,No phone service,DSL,Yes,No,Yes,Yes,No,No,One year,No,Bank transfer (automatic),42.30,1840.75,No
4,9237-HQITU,Female,0,No,No,2,Yes,No,Fiber optic,No,No,No,No,No,No,Month-to-month,Yes,Electronic check,70.70,151.65,Yes


### Observation

The dataset contains customer demographic information, account information, subscribed services, and a target variable named **Churn**.

The target variable appears to indicate whether a customer left the company.

## Dataset Dimensions

In [9]:
df.shape

(7043, 21)

### Observation

The dataset contains 7043 customer records and 21 features.

This dataset size is sufficient for training and evaluating several machine learning models.

## Available Features

In [10]:
df.columns.tolist()

['customerID',
 'gender',
 'SeniorCitizen',
 'Partner',
 'Dependents',
 'tenure',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod',
 'MonthlyCharges',
 'TotalCharges',
 'Churn']

### Observation

The dataset includes:

- Demographic features
- Subscription information
- Billing information
- Service usage information
- Churn status

These variables may influence customer retention behavior.

## Data Types

In [11]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


### Observation

Most variables are categorical and stored as object types.

A numerical variable such as:

- tenure
- MonthlyCharges

are already numeric.

Further investigation is required to verify that all columns have appropriate data types.

## Numerical Statistics

In [12]:
df.describe()

,SeniorCitizen,tenure,MonthlyCharges
count,7043.000000,7043.000000,7043.000000
mean,0.162147,32.371149,64.761692
std,0.368612,24.559481,30.090047
min,0.000000,0.000000,18.250000
25%,0.000000,9.000000,35.500000
50%,0.000000,29.000000,70.350000
75%,0.000000,55.000000,89.850000
max,1.000000,72.000000,118.750000


### Observation

The statistical summary provides information regarding:

- Minimum values
- Maximum values
- Mean
- Standard deviation

The tenure variable shows substantial variation among customers, suggesting different levels of customer loyalty.

## Missing Values Analysis

In [13]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

### Observation

No explicit missing values were detected.

However, some columns may contain blank strings that are not recognized as null values.

Additional investigation is required.

## Duplicate Records Analysis

In [14]:
df.duplicated().sum()

np.int64(0)

### Observation

No duplicate records were found.

This suggests that each row represents a unique customer.

## Target Variable Analysis

In [17]:
df["Churn"].value_counts()

Churn
No     5174
Yes    1869
Name: count, dtype: int64

In [18]:
df["Churn"].value_counts(normalize=True)*100

Churn
No     73.463013
Yes    26.536987
Name: proportion, dtype: float64

### Observation

The target variable contains two classes:

- Yes
- No


## Cardinality Analysis

In [19]:
for col in df.columns:
    print(f"{col}: {df[col].nunique()}")

customerID: 7043
gender: 2
SeniorCitizen: 2
Partner: 2
Dependents: 2
tenure: 73
PhoneService: 2
MultipleLines: 3
InternetService: 3
OnlineSecurity: 3
OnlineBackup: 3
DeviceProtection: 3
TechSupport: 3
StreamingTV: 3
StreamingMovies: 3
Contract: 3
PaperlessBilling: 2
PaymentMethod: 4
MonthlyCharges: 1585
TotalCharges: 6531
Churn: 2


### Observation

Some features contain only a few categories and are suitable for categorical encoding.

Other variables such as customerID contain a unique value for each customer and may not contribute to predictive performance.

## Potential Data Quality Issues

Based on the initial investigation, the following issues should be examined further:

1. Possible incorrect data types.
2. Presence of blank values.
3. Potentially useless identifiers such as customerID.
4. Class imbalance in the target variable.
5. Need for categorical encoding.

# Conclusion

The dataset was successfully loaded and inspected.

Key findings include:

- Identification of the target variable.
- Understanding of feature categories.
- Verification of dataset dimensions.
- Initial assessment of missing values and duplicates.
- Detection of potential preprocessing requirements.

The next step will be Exploratory Data Analysis (EDA), where relationships between features and customer churn will be investigated.